donkeycar/parts/controller.py

```diff
@@ -227,107 +228,138 @@ def poll(self):
-        :param tick: current tick in mu s
-        :return: difference in ticks
-        """
+        self.rc1.set_mode(self.RC1_in_PIN, pigpio.INPUT)
+        self.cb1 = self.rc1.callback(self.RC1_in_PIN, pigpio.EITHER_EDGE, self.cbf)
+        self.rc2.set_mode(self.RC2_in_PIN, pigpio.INPUT)
+        self.cb2 = self.rc2.callback(self.RC2_in_PIN, pigpio.EITHER_EDGE, self.cbf)
+        self.rc3.set_mode(self.RC3_in_PIN, pigpio.INPUT)
+        self.cb3 = self.rc3.callback(self.RC3_in_PIN, pigpio.EITHER_EDGE, self.cbf)
+
+        # print('RCReceiver gpio ' + str(self.RC1_in_PIN) + ' created')
+        # print('RCReceiver gpio ' + str(self.RC2_in_PIN) + ' created')
+        # print('RCReceiver gpio ' + str(self.RC3_in_PIN) + ' created')
+
+    def cbf(self, gpio, level, tick):
         import pigpio
-        if self.high_tick is not None:
-            t = pigpio.tickDiff(self.high_tick, tick)
-            return t
-
-    def _cbf(self, gpio, level, tick):
         """ Callback function for pigpio interrupt gpio. Signature is determined
             by pigpiod library. This function is called every time the gpio
-            changes state as we specified EITHER_EDGE.
+            changes state as we specified EITHER_EDGE.  The pigpio callback library
+            sends the user-defined callback function three parameters, which it may or may not use
         :param gpio: gpio to listen for state changes
         :param level: rising/falling edge
         :param tick: # of mu s since boot, 32 bit int
         """
-        if level == 1:
-            self.period = self._update_param(tick)
-            self.high_tick = tick
-        elif level == 0:
-            self.high = self._update_param(tick)
-
-    def pulse_width(self):
+        if gpio == self.RC1_in_PIN:
+            if level == 1:
+                self.high_tick1 = tick
+            elif level == 0:
+                if self.high_tick1 is not None:
+                    self.high1 = pigpio.tickDiff(self.high_tick1, tick)
+        if gpio == self.RC2_in_PIN:
+            if level == 1:
+                self.high_tick2 = tick
+            elif level == 0:
+                if self.high_tick2 is not None:
+                    self.high2 = pigpio.tickDiff(self.high_tick2, tick)
+        if gpio == self.RC3_in_PIN:
+            if level == 1:
+                self.high_tick3 = tick
+            elif level == 0:
+                if self.high_tick3 is not None:
+                    self.high3 = pigpio.tickDiff(self.high_tick3, tick)
+
+
+    def pulse_width(self, high):
         """
         :return: the PWM pulse width in microseconds.
         """
-        if self.high is not None:
-            return self.high
+        if high is not None:
+            return high
         else:
             return 0.0
 
-    def run(self):
+    def run(self, img_arr=None):
         """
         Donkey parts interface, returns pulse mapped into [MIN_OUT,MAX_OUT] or
         [MAX_OUT,MIN_OUT]
```

Comment:
Can you update the doc string so it really describes what it is doing now?

Reply:
To be honest, I don't know that that string does but I apparently need it or I get an error :-(

It's harmless